[Mellanox]Resolve chassis broken due to inconsistent with latest sonic_platform_common

stephenxs opened the pull request:
**- What I did**
Resolve chassis broken by supporting the latest component API (get_firmware_version) 
Currently get_firmware_version implemented by using chassis.get_firmware_version and chassis._component_name_list which are not supported in the latest sonic_platform_common, causing chassis broken.
Update this part so that it aligns to the latest sonic_platform_common

**- How I did it**
Originally get_firmware_version is a member function of chassis but recently it is moved to component.py. In this case mlnx-platform-api should be updated accordingly otherwise chassis is broken.

**- How to verify it**
test
- whether firmware versions can be fetch
- whether chassis can be loaded successfully

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

Commit:
[sonic_platform]Support component API (get_firmware_version).
Currently get_firmware_version implementated by using chassis.get_firmware_version and chassis._component_name_list which are not supported in the latest sonic_platform_common, causing chassis broken.
Update this part so that it aligns to the latest sonic_platform_common
Support component API

keboliu commented:
in PEP8, usually, it requires two lines between functions definition or classes, would you please go through the files? I would like to suggest to stick with PEP8 formats.

Commit:
[sonic_platform]follow the PEP8 standard

liat-grozovik reviewed (changes requested):
Why do we have asic fw part of platform API
IMO asic fw is not part of the platform and it requires SDK bundle.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,199 @@
+
+#components definitions
+COMPONENT_BIOS = "BIOS"
+COMPONENT_FIRMWARE = "ASIC-FIRMWARE"
```
@stephenxs component ASIC

stephenxs replied:
this part has been remvoed.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,199 @@
+        return cpld_version
+
+
+class ComponentASIC_FW(Component):
```
@stephenxs component ASIC

stephenxs replied:
this part has been remvoed.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/platform.py:
```diff
@@ -18,12 +18,12 @@ def __init__(self):
         PlatformBase.__init__(self)
         if self._is_host():
             self._chassis = Chassis()
+            self._chassis.initialize_components_list()
```
@stephenxs initialize_components() or initialize_component()

stephenxs replied:
fixed

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,199 @@
+FW_QUERY_VERSION_COMMAND = 'mlxfwmanager --query'
+BIOS_QUERY_VERSION_COMMAND = 'dmidecode -t 11'
+CPLD_VERSION_FILE_PATTERN = '/var/run/hw-management/system/cpld?_version'
+CPLD_VERSION_MAX_LENGTH = 4
```
@stephenxs what if platform has 5, 3 or 2 CPLDs? How you will distinguish?

stephenxs replied:
This is the max length of version file. Currently the max version length is two which means it contains 2 digit at last. Anyway I think using four digits to represent the version number is enough.
It's not the number of CPLDs.

nazariig replied:
@stephenxs ok. Thanks.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,199 @@
+
+FW_QUERY_VERSION_COMMAND = 'mlxfwmanager --query'
+BIOS_QUERY_VERSION_COMMAND = 'dmidecode -t 11'
+CPLD_VERSION_FILE_PATTERN = '/var/run/hw-management/system/cpld?_version'
```
@stephenxs i think this is inaccurate.

stephenxs replied:
we can use cpld[0-9]_version

stephenxs replied:
fixed.

Commit:
[sonic_platform]Address comments
1. rename initialize_components
2. replace cpld?_version with cpld[0-9]_version in order to be more accurate

Commit:
[sonic_platform]Remove ASIC_FW from components list

Commit:
[sonic_platform]Add description of components

stephenxs commented:
> Why do we have asic fw part of platform API
> IMO asic fw is not part of the platform and it requires SDK bundle.

ASIC FW part has been removed.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,144 @@
+        cpld_version_file_list = glob(CPLD_VERSION_FILE_PATTERN)
+        cpld_version = ''
+        if cpld_version_file_list is not None and not cpld_version_file_list == []:
+            cpld_version_file_list.sort()
```
@stephenxs this logic does not cover the generic case: when the symbolic links do not contain numbers. The result of the aggregated CPLD version (cpld1.cpld2.cpld3, etc.) will be invalid. Eventually this may cause unexpected behaviour which will be hard to catch/debug.

stephenxs replied:
IMO
1. it's hw-management's obligation to ensure the symbolic links follow the correct name convention.
2. Currently those cpld version files are symbolic links to the files in sysfs. And the sysfs file hierarchy is determined by hardware, to be more precise, something like CPLDs' bus addresses which are very less likely to be changed among variant versions of hw-mgmt or CPLD firmware.

nazariig replied:
>     1. it's hw-management's obligation to ensure the symbolic links follow the correct name convention.

@stephenxs ok. We can try rely on that.

stephenxs commented:
Do you mean something like “cpld_version” or other name? IMO, hw-management should make sure the symbol link names follow the correct name convention.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,144 @@
+        """
+        cpld_version_file_list = glob(CPLD_VERSION_FILE_PATTERN)
+        cpld_version = ''
+        if cpld_version_file_list is not None and not cpld_version_file_list == []:
```
@stephenxs in case if cpld_version_file_list is empty we might need to throw the exception.

stephenxs replied:
fixed.

nazariig replied:
Maybe something like this will be better:
```
if cpld_version_file_list is not None and not cpld_version_file_list.empty():
```

stephenxs replied:
According to PEP8, the recommended practice is:

    if cpld_version_file_list is not None and cpld_version_file_list:

We should go this way.

Commit:
[sonic_platform] Raise an exception if CPLD version files not found.

nazariig commented:
> 
> Do you mean something like “cpld_version” or other name? IMO, hw-management should make sure the symbol link names follow the correct name convention.

@stephenxs yes. This may be a case.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,146 @@
+        """
+        result = ''
+        try:
+            fileobj = io.open(filename)
```
@stephenxs consider using "with" statement

stephenxs replied:
fixed

Commit:
[sonic_platform]Address comments
1. implement file reading by with statment
2. adjust statement which testing whether list is empty

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,145 @@
+        return self.name
+
+
+    def _read_generic_file(self, filename, len):
```
@stephenxs suggest to move this to a separate helper class so later on this can be reused.

stephenxs commented:
Yes. But this is out of the scope of this PR. We have a backlog in which all common functions will be extracted and shared by all of platform APIs.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,145 @@
+            result = stdout.rstrip('\n')
+
+        except OSError, e:
+            result = ''
```
@stephenxs how do you suppose to handle errors here? IMHO, this is critical so the exception should be handled in a proper way by upper layer. Also the same comment as above: consider moving to a separate helper class/namespace because this has nothing to do with the components.

stephenxs replied:
Having reflected it, I think upper layer of platform api is platform-independent and not supposed to know and handle platform-dependent details, like variant kinds of exceptions because such behavior seems to violate the pricinple of software layering and modularization.
But what you said is also reasonable, upper layer should be notified in case of this kind of errors. To achieve that, we have some options:
1. platform api returns a certain value indicating an error occurs, for example an empty string or None. upper layer test the return value and handle the error case.
2. platform api catch the platform-specific exceptions and raise a runtime error that will be handled by upper layer.

I prefer the second option since it's obvious an error here.
Meanwhile, I also suggest having platform api clearly defined regarding error return and handling.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,145 @@
+        try:
+            m = re.search('OEM[\s]*Strings\n[\s]*String[\s]*1:[\s]*([0-9a-zA-Z_\.]*)', bios_ver_str)
+            result = m.group(1)
+        except:
```
@stephenxs IMHO, if we fail to return the valid version string we should throw the exception unless other behaviour is not defined by platform API.

nazariig commented:
> 
> Yes. But this is out of the scope of this PR. We have a backlog in which all common functions will be extracted and shared by all of platform APIs.

@stephenxs ok.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -0,0 +1,145 @@
+            with io.open(filename, 'r') as fileobj:
+                result = fileobj.read(len)
+            return result
+        except Exception as e:
```
@stephenxs IMHO, this is critical so the exception should be handled in a proper way by upper layer.

Commit:
[sonic_platform]Raise exception if any error occurs

stephenxs commented:
Community PR opened [[Mellanox]Resolve chassis broken due to inconsistent with latest sonic_platform_common #3569](https://github.com/Azure/sonic-buildimage/pull/3569)
